### PR TITLE
*: Transition from tap Diagnostic(...) to YAML(...)

### PR DESCRIPTION
**Pull request by wking**
See ad47e7d5 (#439) about how node-tap handles YAML blocks vs. diagnostics.

### CODE REVIEW

**Comment by Mashimiao** [comment]
Please use `gofmt -s -w validation/create.go` to fix CI failure

**Comment by wking** [comment]
On Sun, Dec 03, 2017 at 09:26:25PM -0800, Ma Shimiao wrote:
> Please use `gofmt -s -w validation/create.go` to fix CI failure

Oops, thanks.  Done with 206845a → e666abd.

**Comment by Mashimiao** [comment]
The result doesn't looks the same as result in README.md
```
# make RUNTIME=runc localvalidation
RUNTIME=runc tap validation/root_readonly_true.t validation/process_rlimits.t validation/linux_rootfs_propagation_unbindable.t validation/linux_readonly_paths.t validation/linux_masked_paths.t validation/mounts.t validation/process.t validation/root_readonly_false.t validation/linux_sysctl.t validation/linux_devices.t validation/linux_id_mappings.t validation/process_capabilities.t validation/process_oom_score_adj.t validation/create.t validation/hostname.t validation/linux_rootfs_propagation_shared.t validation/default.t
total ................................................... 0/0

ok
```

Forgot to modify Makefile?

**Comment by wking** [comment]
On Sun, Dec 03, 2017 at 09:57:08PM -0800, Ma Shimiao wrote:
> # make RUNTIME=runc localvalidation

The README uses ‘sudo’, but I'll assume your ‘#’ prompt means you're
running as root.

> RUNTIME=runc tap validation/root_readonly_true.t … validation/default.t
> total ................................................... 0/0

That is very strange.  I'm not able to reproduce that myself.  Is it
actually node-tap?  Is it executing all of the listed test
executables?

> Forgot to modify Makefile?

This PR should not require any Makefile updates.

**Comment by Mashimiao** [comment]
So strange, The following is my environment:
```
# which tap
/usr/bin/tap
# rpm -qf /usr/bin/tap
nodejs-tap-0.4.4-9.fc24.noarch
# file /usr/bin/tap
/usr/bin/tap: symbolic link to ../lib/node_modules/tap/bin/tap.js
# uname -r
4.13.13-200.fc26.x86_64
```

**Comment by Mashimiao** [comment]
And
```
# ./validation/create.t 
TAP version 13
ok 1 - create MUST generate an error if the ID is not provided
  ---
  {
    "error": "exit status 1",
    "reference": "https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#create",
    "stderr": "/usr/local/sbin/runc: \"create\" requires exactly 1 argument(s)\n"
  }
  ...
ok 2 - create MUST create a new container
  ---
  {
    "reference": "https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#create"
  }
  ...
ok 3 - 
```
I think that's not what we expect

**Comment by wking** [comment]
On Sun, Dec 03, 2017 at 10:37:46PM -0800, Ma Shimiao wrote:
> # rpm -qf /usr/bin/tap
> nodejs-tap-0.4.4-9.fc24.noarch

I haven't looked too closely, but it seems like nodejs-tap improved
it's YAML handling in 1.0.0 [1].  You may want to test with a more
recent version (I'm using the current 11.0.0).  If the issue is
nodejs-tap, you can also isolate the problem by using a script that
just echos some canned TAP:

  $ cat test-yaml 
  #!/bin/sh
  cat <<EOF
  TAP version 13
  1..4
  ok 1 - success diagnostic
  # success
  not ok 2 - failure diagnostic
  # failure
  ok 3 - success YAML
    ---
    message: success
    ...
  not ok 4 - failure YAML
    ---
    message: failure
    ...
  EOF
  $ tap ./test-yaml
  ./test-yaml ........................................... 2/4
    not ok failure diagnostic
    not ok failure YAML
      message: failure

  total ................................................. 2/4


    2 passing (40.267ms)
    2 failing

[1]: https://github.com/tapjs/node-tap/issues/109#issuecomment-101390385

**Comment by wking** [comment]
On Sun, Dec 03, 2017 at 10:54:09PM -0800, Ma Shimiao wrote:
> ok 1 - create MUST generate an error if the ID is not provided
>   ---
>   {
>     "error": "exit status 1",
>     "reference": "https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#create",
>     "stderr": "/usr/local/sbin/runc: \"create\" requires exactly 1 argument(s)\n"
>   }
>   ...

This is what we expect.  runc errored, and we expected an error, so
the test passes.

> ok 3 - 

We expect a passing test here, but having a test description would
help :p.  I've added one with e666abd → a55d234, which also fixes the
fourth create.t test so it hits the runtime.

But I don't think either of those would be the source of your 0/0
results.

**Comment by wking** [comment]
Ah, possible cause for your 0/0 is that you didn't [run][1]:

```console
$ make runtimetest validation-executables
```

first.  I've added a guard for that to the `Makefile` with a55d234 → 07a3809.

[1]: https://github.com/opencontainers/runtime-tools/blame/efaa6692f71c00e0525a4deb0e152af63e1faea8/README.md#L41

**Comment by Mashimiao** [comment]
No, I did run `$ make runtimetest validation-executables`.

**Comment by wking** [comment]
Can you `strace -f ... tap validation/mounts.t` to see if `tap` is actually executing `mounts.t`?

**Comment by wking** [comment]
This is a followup to #439.  Can we add it to [the 0.4.0 milestone][1] and land it before #308?  I'd rather switch #308 over to the `.YAML(…)` method too.

[1]: https://github.com/opencontainers/runtime-tools/milestone/6

**Comment by q384566678** [comment]
@Mashimiao @mrunalp @liangchenye What do you think of this pr?

File: README.md [hunk 1]
```diff
@@ -39,15 +39,18 @@ $ npm install tap
 
 ```console
 $ make runtimetest validation-executables
+$ sudo make RUNTIME=runc localvalidation
```
Comment (q384566678) [inline README.md]: Do we need to add the following? 

``
for EXECUTABLE in runtimetest validation/linux_rootfs_propagation_shared.t validation/create.t validation/linux_gid_mappings.t validation/linux_rootfs_propagation_unbindable.t validation/linux_readonly_paths.t validation/linux_masked_paths.t validation/mounts.t validation/process.t validation/root_readonly_false.t validation/linux_sysctl.t validation/default.t validation/linux_devices.t validation/process_capabilities.t validation/process_oom_score_adj.t validation/process_rlimits.t validation/root_readonly_true.t validation/hostname.t validation/linux_uid_mappings.t; \
do \
	test -x "${EXECUTABLE}"; \
done || (echo "missing test executable; run 'make runtimetest validation-executables'" >&2 && exit 1)
``

The other looks good to me.

Comment (wking) [reply README.md]: > Do we need to add the following?
>
>     for EXECUTABLE
>     …

I've pushed 07a3809 → 23f547a to remove that noise from the `localvalidation` output.

**Comment by q384566678** [comment]
LGTM

**Comment by q384566678** [comment]
ping @liangchenye @Mashimiao

**Comment by q384566678** [comment]
This is an important pr that progressed to the 0.4 version, WDYT @liangchenye .

**Comment by liangchenye** [comment]
It looks good to me and sorry @wking I merged 547 just now, so we have a conflict to solve.

Since the runtime-CLI interface is added now. https://github.com/opencontainers/runtime-tools/pull/321
we need to change the 'create' function a lot. But we can do that in our 0.5 milestone.

**Comment by wking** [comment]
Rebased onto master with 23f547a → e85081a.

**Comment by q384566678** [comment]
LGTM

**Comment by liangchenye** [comment]
LGTM

**Pull request merged**